Add qwen-code support, Free credits

nitinprajwal opened the pull request:
### Related GitHub Discussion

Implements: Discussion #6861 - "Support Qwen Code CLI as an API Provider"
Link: https://github.com/RooCodeInc/Roo-Code/discussions/6861


### Description

This PR implements support for Qwen Code CLI as an API provider in Roo Code, addressing the community feature request from GitHub Discussion #6861.

**Key Implementation Details:**

- **OAuth Authentication**: Implements OAuth2 flow for Qwen API authentication using cached credentials
- **Provider Integration**: Complete integration following existing provider patterns
- **UI Components**: Added settings UI for OAuth path configuration
- **Error Handling**: Comprehensive error handling with proper user feedback
- **Token Management**: Automatic token refresh and validation

**Architecture Decisions:**

- Used OpenAI-compatible API format for consistency with existing patterns
- Added configurable OAuth credentials path for flexibility
- Followed existing provider schema patterns for seamless integration

**Files Modified:**

- `src/api/providers/qwen-code.ts` - Main provider implementation
- `packages/types/src/providers/qwen-code.ts` - Type definitions and model configurations
- `webview-ui/src/components/settings/providers/QwenCode.tsx` - Settings UI component
- `packages/types/src/provider-settings.ts` - Provider schema integration
- `src/api/index.ts` - Provider registration

### Test Procedure

**Manual Testing Steps:**

1. **Setup OAuth Credentials:**

    - Obtain Qwen OAuth credentials and place in `~/.qwen/oauth_creds.json`
    - Or configure custom path in settings

2. **Provider Configuration:**

    - Open Roo Code settings
    - Select "Qwen Code" as API provider
    - Verify OAuth path configuration UI appears
    - Test with both default and custom OAuth paths

3. **Functionality Testing:**

    - Create a new conversation
    - Send test prompts to verify streaming responses
    - Test error handling with invalid credentials
    - Verify token refresh functionality

4. **Integration Testing:**
    - Test alongside other providers
    - Verify provider switching works correctly
    - Check that model selection UI shows Qwen models
    
### Screenshot
<img width="561" height="802" alt="image" src="https://github.com/user-attachments/assets/c22b886b-47af-4af3-8c62-3b480565f9be" />


**Expected Results:**

- Qwen Code provider appears in provider list
- OAuth authentication works seamlessly
- Error messages are user-friendly and actionable

### Pre-Submission Checklist

- [x] **Discussion Referenced**: This PR implements the feature request from GitHub Discussion #6861
- [x] **Scope**: Changes are focused on Qwen Code provider implementation
- [x] **Self-Review**: Performed thorough self-review of all code changes
- [x] **Testing**: Manually tested OAuth flow, streaming, and error handling
- [x] **Documentation Impact**: Added inline documentation and user-facing settings
- [x] **Contribution Guidelines**: Read and followed contributor guidelines

**Settings UI:**

- New "Qwen Code" option in provider dropdown
- OAuth credentials path configuration field
- Helpful description text and setup instructions


### Documentation Updates

- [x] No additional documentation updates required
- Inline code documentation provided
- Settings UI includes helpful descriptions and setup guidance
- Error messages provide clear instructions for users

### Additional Notes

**Implementation Highlights:**

- **Robust Error Handling**: Comprehensive error handling for OAuth failures, network issues, and credential problems
- **Token Management**: Automatic token refresh with proper expiry handling
- **Configurable Paths**: Support for custom OAuth credential file locations
- **Type Safety**: Full TypeScript integration with proper type definitions

**Technical Considerations:**

- OAuth credentials are cached locally following Qwen's standard patterns
- Token refresh happens automatically when tokens are near expiry

**Future Enhancements:**

- Could add support for additional Qwen models as they become available
- Potential for enhanced OAuth flow with browser-based authentication
- Possible integration with Qwen's upcoming API features

### Get in Touch

**Discord Username:** @nitinprajwal

Feel free to reach out if you have any questions about the implementation, need clarification.
<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Adds Qwen Code support as an API provider with OAuth2 authentication, UI integration, and comprehensive error handling.
> 
>   - **Behavior**:
>     - Adds Qwen Code as an API provider with OAuth2 authentication in `qwen-code.ts`.
>     - Implements token management with automatic refresh and validation.
>     - Integrates Qwen Code into provider selection and configuration UI in `ApiOptions.tsx` and `QwenCode.tsx`.
>   - **Provider Integration**:
>     - Registers `QwenCodeHandler` in `index.ts` and `providers/index.ts`.
>     - Adds Qwen Code model definitions in `providers/qwen-code.ts`.
>     - Updates `provider-settings.ts` to include Qwen Code in provider schemas.
>   - **UI and Settings**:
>     - Adds Qwen Code settings UI in `QwenCode.tsx`.
>     - Updates `settings.json` for Qwen Code descriptions and instructions.
>     - Modifies `constants.ts` and `useSelectedModel.ts` to support Qwen Code models.
>   - **Validation and Configuration**:
>     - Updates `checkExistApiConfig.ts` and `validate.ts` to handle Qwen Code configuration without API keys.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for ebe1f7711fd60fd79089ae93af844f34e7f8b65e. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

daniel-lxs commented:
Hey @nitinprajwal, do you know if this doesn't break any TOS with Qwen? I couldn't find any information about using this OAuth authentication.

mechanicmuthu reviewed (commented):
Well done.

nitinprajwal commented:
Thank you @mechanicmuthu

nitinprajwal commented:
Hey @daniel-lxs, can you help me with these check-types (it was passing all the checks before), the issue is adding new provider "qwen-code" to the providerNames array, but the @roo-code/cloud doesn't support the new "qwen-code". This caused TypeScript errors when passing ClineProvider instances.

nitinprajwal commented:
Hey @daniel-lxs , thanks for the feedback! I've implemented the custom OAuth path in src/api/providers/qwen-code.ts. This should now align with the configuration field in the UI and provider settings schema.
I also thoroughly tested the changes to ensure the custom path works as expected. Additionally, I rebased the branch to resolve all conflicts and keep everything in sync.

synapticchro commented on src/api/providers/qwen-code.ts:
```diff
@@ -0,0 +1,294 @@
+	}
+
+	private getBaseUrl(creds: QwenOAuthCredentials): string {
+		let baseUrl = creds.resource_url || "https://dashscope.aliyuncs.com/compatible-mode/v1"
```
Yes, it's the same mentioned in the docs

https://www.alibabacloud.com/help/en/model-studio/compatibility-of-openai-with-dashscope

NaccOll replied:
Because Aliyun uses a proxy to simulate the OpenAI-compatible API, accessing it directly will result in a 404 error.

However, ${base}/models and ${base}/completions are available.
